### PR TITLE
Close open WebSocket sessions when application is shutting down

### PR DESCRIPTION
**Pull request by oising**
When a dotnet core process hosting hotchocolate is signaled to shutdown, any open websocket sessions (i.e. subscriptions) will cause the shutdown process to stall until the default timeout, which is 30 seconds. At this point, the process is forcefully terminated. While the default is 30 seconds, some people (namely me, lol) have longer timeouts due to other things going on in their server. This makes my k8s rolling upgrades take a lot longer than I'd like :) 

I noticed that WebSocket sessions are not explicitly closed on application shutdown. This PR adds that functionality.

### CODE REVIEW

File: src/HotChocolate/AspNetCore/src/AspNetCore/WebSocketSubscriptionMiddleware.cs [hunk 1]
```diff
@@ -2,24 +2,28 @@
...
         IRequestExecutorResolver executorResolver,
         IHttpResultSerializer resultSerializer,
         IServerDiagnosticEvents diagnosticEvents,
+        IHostApplicationLifetime hostLifetime,
         NameString schemaName)
         : base(next, executorResolver, resultSerializer, schemaName)
     {
         _diagnosticEvents = diagnosticEvents ??
-            throw new ArgumentNullException(nameof(diagnosticEvents));
+                            throw new ArgumentNullException(nameof(diagnosticEvents));
```
Comment (PascalSenn) [inline src/HotChocolate/AspNetCore/src/AspNetCore/WebSocketSubscriptionMiddleware.cs:25]:
```suggestion
            throw new ArgumentNullException(nameof(diagnosticEvents));
```

File: src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketSession.cs [hunk 2]
```diff
@@ -35,7 +38,8 @@ public class WebSocketSession : ISocketSession
 
     public async Task HandleAsync(CancellationToken cancellationToken)
     {
-        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(
+            new [] { cancellationToken, _applicationStopping });
```
Comment (PascalSenn) [inline src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketSession.cs:42]: I think there exist a overload of CreateLinkedTokenSource with exactly two parameters :)

**Comment by michaelstaib** [comment]
Will have a look at this tomorrow.

**Review by PascalSenn — COMMENTED** [review]
Thanks for the PR! :) 👏 

This looks good to me, @michaelstaib  you have the subscriptions currenlty on the brain stack, any objections for this?

File: src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketSession.cs (hunk 2 above)
Comment (oising) [reply src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketSession.cs:42]: Are you getting alloc allergies? ;) I'll fix it.

Comment (oising) [reply src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketSession.cs]: done

**Comment by michaelstaib** [comment]
Related #4813

**Comment by oising** [comment]
> Will have a look at this tomorrow.

12.7.1 ?

**Review by michaelstaib — APPROVED** [review]
I have integrated it into the new subscription layer.

**Pull request merged**